test/Test/Database/LSMTree/Internal/Snapshot.hs

```diff
@@ -0,0 +1,262 @@
+roundtripCBOR _ = explicitRoundtripCBOR encode decode
+
+-- | See 'explicitRoundtripCBOR'.
+roundtripCBOR' :: (Encode a, DecodeVersioned a, Eq a, Show a) => Proxy a -> a -> Property
+roundtripCBOR' _ = explicitRoundtripCBOR encode (decode' currentSnapshotVersion)
+
+-- | @fromFlatTerm . toFlatTerm = id@
+--
+-- This will also check @validFlatTerm@ on the result of @toFlatTerm@.
+explicitRoundtripFlatTerm ::
+     (Eq a, Show a)
+  => (a -> Encoding)
+  -> (forall s. Decoder s a)
+  -> a
+  -> Property
+explicitRoundtripFlatTerm enc dec x = case back flatTerm of
+    Left e ->
+            counterexample
+              ("Decoding failed: " <> show e)
+              (property False)
+    Right y ->
+           counterexample
+             ("Expected value and roundtripped value do not match")
+             (x === y)
+      .&&. counterexample
+             ("Invalid flat term")
+             (property (validFlatTerm flatTerm))
+  where
+    flatTerm = there x
+
+    there = toFlatTerm . enc
+    back = fromFlatTerm dec
+
+-- | See 'explicitRoundtripFlatTerm'.
+roundtripFlatTerm ::
+     (Encode a, Decode a, Eq a, Show a)
+  => Proxy a
+  -> a
+  -> Property
+roundtripFlatTerm _ = explicitRoundtripFlatTerm encode decode
+
+-- | See 'explicitRoundtripFlatTerm'.
+roundtripFlatTerm' ::
+     (Encode a, DecodeVersioned a, Eq a, Show a)
+  => Proxy a
+  -> a
+  -> Property
+roundtripFlatTerm' _ = explicitRoundtripFlatTerm encode (decode' currentSnapshotVersion)
+
+{-------------------------------------------------------------------------------
+  Arbitrary: versioning
+-------------------------------------------------------------------------------}
+
+instance Arbitrary SnapshotVersion where
+  arbitrary = elements [V0_0]
+  shrink V0_0 = []
+
+deriving newtype instance Arbitrary a => Arbitrary (Versioned a)
+
+{-------------------------------------------------------------------------------
+  Arbitrary: SnapshotMetaData
+-------------------------------------------------------------------------------}
+
+instance Arbitrary SnapshotMetaData where
+  arbitrary = SnapshotMetaData <$> arbitrary <*> arbitrary <*> arbitrary
+  shrink (SnapshotMetaData a b c) =
+      [ SnapshotMetaData a' b' c'
+      | (a', b', c') <- shrink (a, b, c)]
+
+deriving newtype instance Arbitrary SnapshotLabel
+
+instance Arbitrary Text where
+  arbitrary = Text.pack <$> arbitrary
+  shrink x = Text.pack <$> shrink (Text.unpack x)
+
+{-------------------------------------------------------------------------------
+  Arbitrary: TableConfig
+-------------------------------------------------------------------------------}
+
+instance Arbitrary TableConfig where
+  arbitrary =
+          TableConfig <$> arbitrary <*> arbitrary <*> arbitrary
+      <*> arbitrary   <*> arbitrary <*> arbitrary <*> arbitrary
+  shrink (TableConfig a b c d e f g) =
+      [ TableConfig a' b' c' d' e' f' g'
+      | (a', b', c', d', e', f', g') <- shrink (a, b, c, d, e, f, g) ]
+
```

Comment:
These instances are intended to be dumb -- I just wanted to enumerate the constructors, so the distribution of generated values is not smart. Do you still think it makes sense to put these generators elsewhere?

Reply:
Depends on whether we want to re-use them. We currently only use fixed table config in the other tests (or at least mostly fixed in StateMachine tests). So I'd say it's okay to leave them here for now and use different generators in different tests.

It might make sense to sense to cover more config variations in other tests, but that's not specific to this PR.